### ISSUE TITLE
Question for further pre-training on a new downstream task

### ISSUE THREAD

**Issue by h19920918** [google-research/text-to-text-transfer-transformer#163]
Hi, firstly thank you for your dedication about it.

I'm implementing further pre-training the pre-trained weights (provided by this repo.) to a new downstream task.

As I know from the paper, the pre-trained weights was trained from the 'C4' dataset and supervised downstream tasks.
I concern about that whether I get performance improvement after pre-training only contexts of a new task.

Could you give to me any advice?

The answer for my question is that I don't expect any improvement from further pre-training the new task since the pre-trained weights was trained with supervised objective for all downstream tasks mentioned in the paper.

It means that the weights have already the ability for the language and inference, from the pre-training method in the paper.
One of the reason about the above is that the pre-trained weights quite works well on a new task.

Additionally, for the input sequences in the pre-training,  which one is the right result of pre-processing for the masking method?

1.
Input: We hold X to be Y that (EOS)
target: (BOS) X these truths Y self evident Z (EOS)
 
2.
Input: (BOS) We hold X to be Y that (EOS)
target: (BOS) X these truths Y self evident Z (EOS)

And, in the previous issue that I ask to you, you answered using 'dynamic masking' but examples will be seen one time due to the large size of dataset.
Suppose that masked tokens will be changed, if an example is used two or three time in the pre-training process.
But the unique tokens are learned in different token representations for each time step.

Can unique mask tokens learn properly from different tokens?

**Comment by craffel** [comment]
Hi, do you mean further training on a supervised task? It is almost always helpful to do further fine-tuning on a downstream supervised task because the multitask model has difficulty at performing optimally in all tasks at once. You can compare the results of rows 2 and 3 from table 12 in our paper; task-specific fine-tuning clearly helps. Note that for the final results in our paper we did additional task-specific fine-tuning. If you mean further pre-training on an unsupervised task, this can help if the unsupervised task is closer in its domain to the supervised tasks you will eventually fine-tune on.

Regarding your question about (BOS) at the start of the inputs, no, there is no BOS token there (option 1). This is not something you need to worry about though, it is handled by the data infeed pipeline automatically.

I don't understand your question about unique tokens:

> But the unique tokens are learned in different token representations for each time step.
> Can unique mask tokens learn properly from different tokens?

What do you mean by "unique tokens are learned in different token representations"? The mask token IDs will be the same regardless of the example or the number of times the example has been seen.

**Comment by h19920918** [comment]
@craffel Thank you for your answer.

Actually, it means suppose that

The original text is "Thank you for your answer"

In the step 1, (your is selected to mask)
Input: Thank you for (X) answer (EOS)
Target: (BOS) (X) your (Y) (EOS)

And, in the another step N, (for is selected to mask)
Input: Thank you (X) your answer (EOS)
Target: (BOS) (X) for (Y) (EOS)

In this case, (X) learned the representation from the "for" and "your" given sentence.
But in the generation process, given (X) what will be next predicted token?
Because (X) learned to predict "for" and "your" in the steps.


The actual meaning of this question is to further pre-train a new context by the unsupervised objective.
When I further pre-trained on the SQuAD context by the unsupervised objective, it showed significant dropping performance.

I suspected occurring this fact from that the pre-trained weights was already pre-trained from the supervised datasets by own label.
So, it has some inference ability by the pre-training with the multi supervised datasets.

Am I correct?

**Comment by craffel** [comment]
> In this case, (X) learned the representation from the "for" and "your" given sentence.
> But in the generation process, given (X) what will be next predicted token?
> Because (X) learned to predict "for" and "your" in the steps.

The model is not learning that a particular mask token (e.g. (X)) maps to a specific word in general; it just learns that the mask token (X) means "fill in this word" and it performs the in-filling based on surrounding context. The model does not learn to associate the mask tokens with any particular words. The same is true in all models that use an unsupervised masking objective.

> When I further pre-trained on the SQuAD context by the unsupervised objective, it showed significant dropping performance.

You mean that you are pre-traning on SQuAD contexts with the unsupervised objective before fine-tuning on SQuAD? And it is hurting performance? I think the model is probably overfitting to the contexts quickly because it is a tiny dataset and it has already seen this data many times.

**Comment by h19920918** [comment]
@craffel 
For the first question, I agree your opinion.

For the second, 
Yes, I pre-trained on SQuAD contexts with unsupervised objective before fine-tuning on SQuAD.
Given a new task, does it improve performance trained by the above manner?

**Comment by craffel** [comment]
> For the first question, I agree your opinion. It was a stupid question. Sorry.

No stupid questions, no worries.

> Yes, I pre-trained on SQuAD contexts with unsupervised objective before fine-tuning on SQuAD.
Given a new task, does it improve performance trained by the above manner?

I think people have shown that this kind of intermediate unsupervised pre-training on the downstream data that will eventually be used for fine-tuning can be helpful. However, this is usually applied to models that have only been trained on unlabeled text data before being fine-tuned on a supervised task. Since T5 was already trained on supervised SQuAD data, this method may not be applicable here.
